### ISSUE TITLE
Improve Prioritisation and Task Allocation in Godot's Development

### ISSUE THREAD

**Issue by WickedInsignia** [godotengine/godot-proposals#8523]
### Describe the project you are working on

I'm a professional Art Director working on AAA games and a regular Godot contributor through sample scenes, bug reports and proposals.

### Describe the problem or limitation you are having in your project

Godot has a "work on what you like" management structure, as detailed [HERE](https://x.com/reduzio/status/1569263960607580160?s=20).
This is nice on paper: people work on what they like and aren't designated onto things that they don't want to do.

Unfortunately this is not a historically productive management style.
The reason executive management exists (and should not be seen as something solely applicable to the corporate world) is because people further down the chain need to be able to focus in on their specific roles. They cannot maintain the perspective and responsibilities of those further up the chain. An engineer can work on whatever they want, but without a lead guiding their tasking it's completely dubious whether their efforts will be spent on things that really need it.

This objectively unfounded management ideal is affecting Godot in some negative ways.
Many features seem to be implemented only so far as the "fun work" is done, with the crucial and tedious work to ensure those features are practical regularly left to the wayside. There is some work that is simply not desirable and never will be, but needs to be done nonetheless to make for a better engine.
The lightmapper, realtime GI, asset workflow, visual shader, advanced import dialogue, material allocation UX and more are all functional on a surface level and probably perfectly acceptable to the engineer that built them. When compared to other engines and the needs of artists however, they have major shortcomings and aren't practical for anything beyond the most trivial of games.
This is compounded by a lack of dogfooding in the core development team (at least to the level usually expected of successful software products, and other remarkable FOSS projects such as Blender).

Many crucial bugs on core systems have been ignored for years. I myself have been intensively testing the graphics side of Godot and identified core issues with many of the lighting features, with many proposals and reports initially discussed or met with approval and then left to sit for a year or more untouched.
Rather than a refinement of the current features, I'm regularly watching entirely new ones be implemented or the renderer rewritten for the umpteenth time, promising a different system that's more efficient in some areas but has the same number of issues in different places than before.
Godot feels like it's in a constant pre-alpha, with no "v1.0" achieved to justify the momentum towards v2.0.

It might be expected that users help contribute to patch these areas, but that's not materializing in many cases.
More often than not I've seen features go ignored because the only person on the core dev team who understands how it works is preoccupied with other less crucial feature additions, rather than returning to refine their half-baked implementation.
The experienced game devs who could test features more intensively simply do not interact with the engine because they come along, realize some core feature is broken in a way that completely obstructs their project, they see the PRs for this issue untouched for years, and then they move on to another engine.

This makes it difficult to justify donations or recommendations.
If the paid core development team is only ever prioritizing what interests them and not doing the drudgery required to keep the engine stable and reliable, myself and others cannot confidently recommend the engine as a refined project where devs can expect robust and complete features. Instead, there's always a disclaimer that most features are considered to a surface-deep level, hardly tested and will most likely have many bugs that need to be worked around. If they can't be worked around, you'll likely be waiting until Godot 5.5 for a fix.

Blender is an excellent example of a FOSS project that has succeeded due to sound managerial decision-making. Godot is constantly held in comparison to it for good reason: it's quickly approaching a similar scale of operation.
The difference is that the Blender Foundation maintains a prioritization structure that contributors clearly benefit from. Paid devs don't simply work on what they like: tedious refinement of the core workings is constantly undertaken through prioritization and allocation. Those free contributors who would prefer to simply do what they like know what most needs doing, and can pick from the pile of crucial tasks or proposals.
The lead devs regularly set a good example by working on boring but important features rather than unimportant moonshot rendering experiments.

Some of Godot's most talented contributors have left the project inorganically. I don't believe Godot's perceived directionless development plays any small part in this, and in talking to existing contributors it's evident that some feel they're simply left waiting for crumbs of info or core changes to be made in other areas they simply cannot predict.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I understand the developers may wish to maintain Godot's "work on what you like" managerial structure and reject a more formal designation-based structure entirely. I believe this is a mistake, and not what people contribute to Godot in the hopes for.
In that situation however, it would help to consider the following to attract confidence from contributors:

- **Maintain an official and well-updated roadmap of upcoming features and core development work in all aspects of the engine.** I understand some of this is being communicated in blog posts, but development updates with vague timelines and teasers of upcoming features are not a roadmap. There should be an official roadmap available in one place, comprehensively outlining the intended features for upcoming versions and the priorities of the core dev team in detail.
- **Set a schedule for major feature implementations and tasks.** It's obvious that some major reworkings on core features are required and although they are promised, the timeline is completely organic and commonly set back at a whim (such as core rendering restructures and SDFGI.) This causes serious issues for developers waiting on fixes or feature implementations that their game hinges on. A comprehensive schedule that the core team puts every effort into abiding by would help establish confidence that issues are being worked on, and any pushbacks are held accountable with sound reasoning.
- **Prioritize proposals and features using public tags or lists.** Godot's Github is full to the brim with untouched bug reports and proposals, some for many years. This is a red flag for devs and studios assessing Godot for their own use. Sometimes bugs or proposals are given Milestone tags, but are otherwise completely devoid of prioritization. This makes it difficult for contributors to know what they should be working on, rather than chipping away at non-crucial areas of the engine. Users also have no idea when showstopping bugs or issues will be tackled, and decide it's not worth waiting for them.

This may all seem somewhat corporate in nature, but it's simply good project management practice. Any utilization in corporate settings is due to its efficiency, not because its some trait unique to those settings.
Successful projects do not fall together organically, they utilize these tools to varying degrees to maintain user confidence and spearhead development in a direction that isn't simply "what we felt like at the time."

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

N/A

### If this enhancement will not be used often, can it be worked around with a few lines of script?

N/A

### Is there a reason why this should be core and not an add-on in the asset library?

N/A

**Comment by WickedInsignia** [comment]
Juan clarified within that thread that paid contributors "are generally paid to work on what they like working on and are paid to dedicate more time to it."
[HERE](https://x.com/reduzio/status/1569263963044462594?s=20)

I acknowledge that paid contributors only work withing their realm of expertise, but the idea that they prioritize dedicating time to needed but undesirable work is not what's being expressed by Juan there.
There will be tasks within their realm of expertise that they may not be interested in, but should still be undertaken to ensure features are well-rounded and practical. I don't get the indication that's happening through my own use of the engine and experiences with its development.

There are absolutely no requests in my proposal to force unpaid contributors to work on designated tasks. Only that priority tasks and the overall roadmap are made clearer so that contributors who _do_ want to dig in and do the difficult work know what needs to be done. This is also so that other contributors and users have a good idea of where the engine is headed, what's being prioritized, and how likely a fix for serious issues may be.

**Comment by akien-mga** [comment]
Thanks for the input.

I see there's a lot of misunderstanding about how the Godot project is actually developed nowadays, which I would like to clarify with a well thought out answer, being the lead of the production team.

I need some time to write it, and would prefer the discussion not to derail too much in the meantime, so I'm locking the thread temporarily. Should be unlocked soon for further discussion once I'm done writing my reply.

**Comment by YuriSizov** [comment]
There is a lot to unpack here.

Let me start by saying that we appreciate external feedback on our project management practices. But I then would like to immediately continue by mentioning that you have to be careful with assessments made with an outsider's perspective. While we are wide in the open for the most part of it, a significant part of the management work work goes unseen in private chats and calls, in personal communication, and in our heads.

It is a problem in and of itself that from the outside it is not always clear how the contributors are organized or what the priorities are. We are actively working on this issue, always. Yet at a certain point there are diminishing returns and no amount of communication is going to reach everyone, be understood by everyone. So we prefer to focus on the results of our collective work.

And the results can be seen as grim as you describe, certainly. There is no denying that some things are going slower than you might want, some things get stuck in a limbo. There is no denying that people leave the project unsatisfied sometimes, or never gain enough interest in the first place, intimidated by the amount of work in front of everyone. That is, however, completely normal and not as grim as it may look.

With every release we address over a thousand different things, across the engine, in all areas which currently have maintainers in them. We implement new features but we also fix bugs in existing ones. We don't, and can't, fix all the bugs, but with every release we get to more of them. We strongly believe that each new release of Godot is better than the last one, and that you can make amazing projects with the engine, ship and sell them.

-----

It's all a balancing act. We have to balance new features and bugfixes. We have to balance paid contributors and those who benevolently volunteer their free time. We have to balance spending time on code and communications, promotions. We have to balance importance and validity of each of thousands ideas and issue reports that our vast community has. We have to balance onboarding new contributors to strengthen the team with our own work.

And we have to balance our personal time and lives on top of that. We have to manage our passion to not burn out, but we also want to interact with everyone interested in the project. We also have our own projects, or we work for a company that has its own tasks and goals. Many sponsored contributors work on more than just Godot.

And yes, we want people to feel comfortable contributing and work on things they are interested in and want to work on. And no, paid contributors are organized and not just do whatever. If you've paid close attention to GitHub last month, we've been crunching addressing the regressions reported late into the release cycle. This was done almost exclusively by sponsored contributors, people who are maintainers of the project, who are trusted to work under such pressure and make qualified decisions.

We do work on what we want, but what we want is to make Godot and make Godot better.

-----

Having a clear roadmap for months and years ahead is one of our priorities at the production team. We are aware that you cannot build a project of this scale without this information readily available, even internally. As I've said, there is a lot to balance here, and making it happen is a long-term project, not something that would happen over night or because there is a social pressure for us to do so.

Keep in mind, that Godot has been operating in red for a bit, where the scale was way beyond what the budget was. It's better now and we can get a feel for the ground and get a better footing. We can plan to hire more people to support critical areas, and that doesn't stop at hiring engineers either. This won't happen over night either.

In the mean time, I hope that the blogposts outlining priorities (and presentations from the recent GodotCon) are sufficient to give you an overview of what's going on and what you can expect in the future.

**Comment by akien-mga** [comment]
First of all, thanks for taking the time to write down your thoughts and offering suggestions for improvements.

For the most part, what you outline as what we should be doing _is_ what we're actively working on, and have been doing for years. We know that we are not doing it the best way possible yet, we're aware of things we still want to improve, and we're constantly working on improving them.

We can discuss specifics of what the current state is on each of these points, and how to improve it further, but first we need to clarify the problem statement and the assumptions made about the Godot project.

## Resources

You talk about the "core development team" and "paid developers". Let me clarify what resources Godot has on that front.

We wrote a blog post about this [in July](https://godotengine.org/article/funding-breakdown-and-hiring-process/) listing the developers paid through donations via the SFC (at the time) and now the Godot Foundation. We'll soon have a new blog post to update about the current status after we brought back funding to a sustainable threshold which covers the existing payrolls, and enables us to start thinking about new hires. We're aware that we need to communicate more about this, there's ongoing work on this.

10 developers, some only hired part-time, it's around 7 full-time equivalents (FTE). That's the amount of resources that the Godot Foundation is able to afford. Then W4 also has a number of hires working part or full-time on upstream Godot following project-defined priorities, adding around 3 FTEs.

So that's 10 FTEs who are being paid to work on Godot, and can reasonably be allocated on development priorities identified by the production team, or their respective teams (rendering, core, platforms, etc.).

Out of these, 4 FTEs are dedicated to production work, which includes: Triaging issues and PRs, reviewing and merging PRs (600 per month in average), setting short-term development priorities (notably for the purpose of doing releases), release management itself, documentation, communication, infrastructure, talking to users to identify major pain points, discussing proposals... Yes, not all of those are typical production/management tasks, but with such a small team for such a big project, we have to wear many hats. That limits how much time we can dedicate to major redesigns of our development and management workflows. All of us _on top of that_ also do development in the areas where we are maintainers.

I'm already getting bogged down in numbers so I'll stop here, but I think you can get a feel for how many people the Godot project actually has available to do all the work you'd like us to do. Then contemplate the work we actually do, how many bugs we fix in each release, how we improve the existing features, all while managing the very diverse and conflicting interests of thousands of developers.

TL;DR: There's only so many hours in a day. 40% of the paid staff is dedicated to production/management work and other non-development tasks. And truth be told, we're all working too many hours in average. We can improve the way we organize work and gain some efficiency, but there's no magic wand that would make a team of 10 succeed at doing what requires a team of 25.

## Management philosophy

I think there's a misunderstanding around Juan's 2022 tweets that really needs to be dispelled.

What Juan was trying to communicate is that the majority of people who work on Godot *are* volunteers, and they work on their own priorities. "What they like" is a simplification, it covers both what they personally enjoy, what they technically need for their projects, or what they recognize is a need of the engine users which they want to support. Most of the people volunteering time to contribute to Godot (and I believe that's your case too) do it because they like what Godot is, and want it to get better. They want Godot users to have the best experience they can have with the features that they're using.

Likewise, all paid contractors came to Godot as volunteer contributors, working on what they personally felt was important to them / to what they perceive of the project's needs. They were then hired to be able to dedicate more time to those areas which they were working on, and occasionally start diving into other areas. That's what Juan meant when saying that paid contractors "are generally paid to work on what they like working on and are paid to dedicate more time to it."

It doesn't mean they only work on what gives them immediate gratification and never bother working on less interesting stuff. They work on the areas of the engine which they are personally interested in, and became a maintainer for. Both paid and volunteer contributors go through the bug tracker regularly to fix issues in their areas of expertise. They go through the proposals tracker and discuss proposals, and sometimes implement some (probably not as much as the volume of proposal demands, but again, this is a time/resource issue). They set their priorities based on input from the production team, the community at large, and their experience working on personal projects.

It does however mean that there are areas of the engine where we are bottlenecked by a lack of dedicated maintainer, because no contributor took enough of a "liking" to it to be able to take over maintaining it (either as volunteer, or as a contractor if the project is able to afford them). This *is* a major issue, which we're fully aware of, but again there's only so little we can do with the resources we have available.

As project funding grows, we hope to be able to hire engineers to cover all major areas of the engine, either from the existing pool of contributors who have shown proficiency in a specific area, or external hires of people with experience in these areas.

IMHO, Juan's 2022 tweets that you linked didn't give enough nuance to those statements, and forgot to mention the work of the production team, which at the time in 2022 was mostly just me, with help from Calinou. Now we have a much bigger team of people who focus at least in part on production work (some paid, some volunteers): https://godotengine.org/teams/#production

I fully acknowledge that those tweets could be misinterpreted, and I believe Juan clarified his positions at different times. As a matter of fact, I gave some nuance to this thread myself in 2022: https://twitter.com/Akien/status/1569285398509002753

But what's important is not how one member of the team communicated a specific point they wanted to highlight in 2022, but how we work today, and whether the way we work is adequate for our resources, the needs of the project, and the well-being of our contributors. There are a lot of challenges there, and we're working on it - slowly but surely, while trying to balance all the responsibilities we have to the community.

**Comment by pritish-thub** [comment]
While this may diverge slightly from the current discussion, could you consider reinstating the Godot live stream sessions where the Godot team addressed pressing questions? This would provide the community with valuable insights into the internal workings and keep everyone better informed about ongoing developments.

**Comment by YuriSizov** [comment]
@pritish-thub We are constantly coming back to this idea, but haven't found a productive way to organize it yet. There are way too many developers involved at this point to have a generalized stream like before. So we probably want to have something more targeted, with a pre-established subject and contributors who are actually relevant to discuss it. But it takes more effort to organize at the moment.

Rest assured, we liked that interactivity and would love to do it again!

**Comment by KoxaKoxama** [comment]
> Oops, didn't want to one-up Remi with this wall of text; was in the middle of writing it

Why did you even decided to write this in the first place? Remi is a project manager, but what do you have to do with it?

**Comment by akien-mga** [comment]
Trolls were blocked, reopening.

**Comment by WickedInsignia** [comment]
The concerns expressed in this proposal aren't really regarding any developer's social media. I believe Godot in general is talked about sufficiently, updates and teasers are constantly pushed through blog posts etc.
It also has nothing to do with who the hired devs are, the core team is outlined on the website.

My proposal is outlining the issues that a lack of public-facing prioritization and allocation are creating for contributors and the public perception of Godot's progress. Even if the team isn't simply "doing what they like" the project still presents as haphazard from the outside. Comprehensive public-facing prioritization and roadmapping (beyond milestones that will simply be pushed back to 5.x if incomplete) would help alleviate that.

The issue wasn't just Juan's post: it's that Juan's post is how development appears. I used it as an example because it aptly reflected my experiences.
I'm also not sure what to make of Juan expressing that Godot does not allocate and runs no risk of stretching resources too thin, while Remi is stating above that the team _does_ allocate, is stretched very thin and is running in the red.
Is the advice that people should simply not listen to what Juan says anywhere?

**Comment by reduz** [comment]
@WickedInsignia I think you are just mixing up two things, so here are facts on how I see them:

1. The largest part of Godot contribution is volunteer effort or external entities. This the project has no control of.
2. Managing this for the Godot project is extremely difficult due to the very high volume of contribution.
3. Large parts of the donations are used for hires that work and aid in project management to help this happen, otherwise basically nothing happens, nothing gets reviewed or merged. Nobody gets feedback on their PRs, etc. This is necessary due to the very high volume of volunteer contribution that needs to be organized.
4. The rest of the hires are majorly very skilled generalists who understand and can work on _most_ areas of the engine. 
5. Not all areas of the engine are 100% covered by those hired by the foundation, many still rely on external contribution. The most obvious example is rendering, which is driven mostly by people not hired by the foundation, but it is also the case of other areas such as  physics, asset IO, or C# support.

These later ones I just mentioned are the areas with most "demand for a roadmap" (basically the areas people wants more work put into), coincidentally these are the ones not currently (or not entirely)  covered by the foundation (and also  coincidentally these are the areas you complain the most about, so go figure). So, how is the project going to put up a roadmap or prioritize tasks it currently has nobody hired to do this?

In any case, as the project gets more funds, this will start to change, but you have to think that until two months this was not the case. Hiring takes time so expecting everything to be be solved  in two months is a bit rushed from your part. 

I think things are getting there, and there is still very good contribution going on on the areas you care about (which mostly won´t be road-mapped for now because it's not done by the foundation) but it takes time. Just be patient and don't push people to rush things.

**Comment by reduz** [comment]
Oh also by the way, regarding road maps, the closest you will find to this is proposals opened by the main contributors, like me @clayjohn and others, where future work is put for discussion with the community. 

We could have a central place linking those, but it would not be a roadmap. More like "things we would like to happen when the resources are there to make them".

**Comment by WickedInsignia** [comment]
The largest part of Godot contribution in raw volume may be external entities, but you still decide the direction of the engine and have implemented many of its core systems, including the rendering systems all contributors build on top of, all the way up to 4.0.

Just because Godot is majorly a volunteer effort does not mean that volunteers have control over the direction of the project as a whole or know what would lead it to success. They don't make executive decisions on the core systems.
Although you may not do the brunt of the trench work you're still accountable to leadership, communication and contributor confidence issues.
Managing all of this **is** difficult. That's why I'm proposing some solutions to help manage it.

Godot's funding didn't begin two months ago, and Godot's core team wasn't implemented two months ago. Your ability to prioritize and manage isn't dependent on the funding you received two months ago and never has been. The issues I'm describing are long-lived and don't suddenly change with more funds.
None of this is to rush anyone either. This is a response to Godot's development trajectory since 3.0.

One thing is objectively true: a comprehensive roadmap and clear prioritization would assist contributor confidence immensely. There is no loss to contributors and users knowing what the direction of the engine is in as much detail as can be provided. It may be an inconvenience for you, and it may be a good deal of work to maintain, but it is an extremely useful tool for us.
It doesn't even need to be rigid: dates can be moved and tasks re-prioritized. All that matters is contributors have a comprehensive reference point, even if it frequently shifts.

If Godot's development is so organic that one cannot be maintained, then **therein lies the issue**. Users cannot rely on anything and contributors can not be sure their contributions even matter, they can't determine where the project is headed beyond "I dunno, features just happen sometimes."
You don't build a game engine through completely random and haphazard contributions at random moments, but you seem to figure that's what's happening and is the ultimate goal. I get the feeling accountability is being offloaded here for the convenience of never having to manage any of it in detail, or put work into exposing that detail to the users helping to build Godot.

A comprehensive roadmap, public-facing task prioritization and schedule for the implementation of major features as I've described would only further improve conditions for contributors, at a little more cost of effort and time to you.
Considering the influx of funding, why not hire a project manager to help implement them?

**Comment by WickedInsignia** [comment]
I should stress: None of this is to impede any contributor's ability to build and implement features at their discretion.
This is intended to outline to contributors the direction and priorities of the core team in detail, the areas and specific tasks most in need of work or less in need, and when users can expect new engine updates or feature implementations (without the pressure of a guaranteed delivery.)
Those who want to help develop crucial areas would know where their efforts are best spent, and those who want to contribute haphazardly would understand why their proposal is outside of current scope and to exercise patience as priorities are tackled first (such as a full rendering rewrite).
All of this at a glance, without pestering team members or sifting through blog posts.

**Comment by WickedInsignia** [comment]
You have historically made major contributions and direct core efforts in some way. You inevitably help decide on hires to the dev team, you manage Godot's core optics online (your Twitter is a potent reference point for many Godot users) and you are still making developments on core rendering infrastructure (such as the SDFGI improvements and a possible GPU renderer).
You steer development direction. That's the point. Even if it's only a suggestion, you have huge sway in the direction of the engine and your efforts have defined it from the ground up. If not then please, point out a single individual that has more influence on the direction of the project than you.

The example you brought up around SDFGI is also my point. If it was clear that there was prioritization on the things that affected users most, through the use of task priorities and a roadmap, there wouldn't be any question as to why SDFGI isn't being worked on.
If you want those people off your back, you clarify the development situation and goals through comprehensive and easily-accessible reference. The work isn't just to their benefit: it's to yours as well.

You asked "priority based on what?" and then literally outlined why some rendering tasks are being prioritized: the most crucially dysfunctional areas of the engine that seriously affect users. You're the technical lead who built the majority of Godot: give what needs prioritization your best guess. It'll be better than most.
I wouldn't care about SDFGI if there was a bug causing stutters in every game that should be tackled first. The point is that **myself and everyone else poking about SDFGI don't know that**. There's no comprehensive prioritization being communicated or easy-grab reference point.

"There is so much to do and fix" isn't a reason not to prioritize. That's actually the reason you do it, or hire someone else to do it.
If Godot is an overwhelming mess you can't fix users will see it that way.

**Comment by groud** [comment]
> One thing is objectively true: a comprehensive roadmap and clear prioritization would assist contributor confidence immensely. There is no loss to contributors and users knowing what the direction of the engine is in as much detail as can be provided. It may be an inconvenience for you, and it may be a good deal of work to maintain, but it is an extremely useful tool for us.

So I am a contributor, not paid by the foundation, and contributing of the spare time if find here and there. What you say here is simply false. I don't need a roadmap to tell me what to do and have "the confidence" to work on what I work on. I am like every other volunteer contributor. I read through the social platforms, see what people need (or mainly what they complain about), then I work on whatever I can to help this. I, of course, try to balance the "what I want to work on" and the "what people need", all of this considering how much time a task will take.

I don't want a roadmap because this is added pressure on our shoulders. Every time we claimed we would do something for a given date (precise or not), and it didn't happen (for whatever reason, but often higher priority tasks first), it has always been a pretext for users to come harass us. So it's always better not to promise anything, to avoid things to backfire. And this is not negotiable, I won't commit in implementing something if I am not paid to implement it, sorry.

Now, I don't mind having a central location where contributors would express what they plan to work is they have the time to do so. Let's call it a "Wishmap". But TBH, that kind of already exists and it's basically what is the proposal repository about.

**Comment by reduz** [comment]
@WickedInsignia But that's the thing. I am indeed working on fixing rendering stuff and many others are too, but none of this work is financed by donations to the Godot Foundation, so nobody working on this wants the obligation or pressure to tell publicly what they are working on and when it will be done (if so). 

So, yes, I am aware of a lot of work going on that is not public yet because Its not financed with donations. As such I can't say what or when will make it in. If you take part in development discussions or maintainer meetings, which are public, you can find out more about this of course. It's not like things are done in secrecy, but those working on it want to do it without being bothered.

**Comment by WickedInsignia** [comment]
@groud That's fine. There are people unlike you who would benefit from a comprehensive reference point and clearer goals to maintain confidence.

This is all proposed with the disclaimer that it takes into account the nature of keeping things flexible and allowing contributors to implement features on a whim, while also understanding why support for those features isn't prioritised or that no one is available to provide it.

Not implementing tools to better track and visualise the project shouldn't be avoided due to the fear of public repercussion. That repercussion is already happening, in the form of users upset with directionless development.

**Comment by reduz** [comment]
@WickedInsignia 

> That repercussion is already happening, in the form of users upset with directionless development.

If I have to choose between having these upset users (who are a very small but vocal minority), and making Godot a place that is unfriendly for people to contribute to, I by far choose the former.

**Comment by WickedInsignia** [comment]
@reduz I'm not talking about tasks kept in secrecy for the benefit of working without pressure, and there will be plenty of tasks not in that state.

Regardless I'm sorry, Juan. A game engine that developers depend on for successful projects requires sacrificing a bit of comfort so they understand the direction the project is taking and what they can rely on. That's the nature of devs interacting with your engine: their projects depend on it. Keeping important developments in limbo directly affects their confidence and willingness to engage with the engine. I can very confidently say this as a AAA dev with professional connections and friends assessing Godot.

If you want to pursue moonshot rendering experiments in your own time, start a private engine project. If you want Godot to remain a hobbyist engine with the potential of scaring away any serious development and contribution, continue on this path.
If you want it to be the Blender of game engines as many people want it to be, then I highly suggest considering this proposal as well as learning from Blender's history.

That's just my own take though.

**Comment by WickedInsignia** [comment]
@reduz None of the suggestions I've mentioned are preclusive to a friendly development environment. They're proposed uniquely to provide a better environment for contributors and users of the engine. I believe I've already described how in exhaustive detail.

**Comment by reduz** [comment]
@WickedInsignia If you are a studio that is investing significant resources to Godot, specially "AAA level stuff", you can hire proper Godot related support from the several companies that are already offering it and working with them. They will fix your bugs, prioritize features and create PRs to upstream those.

If you are a hobbyist developer and you can't afford this, then nobody is in any obligation to fix the or prioritize the bugs you want fixed. Everyone has the best of intentions here, but nobody will do anything for you for free.

What you are asking for is basically that contributors (as in, not foundation hires, people donating their time and money to the project) say what they will be working on publicly and commit themselves to social pressure in exchange of nothing.

This is not going to happen, ever, period.

**Comment by groud** [comment]
> @groud That's fine. There are people unlike you who would benefit from a comprehensive reference point and clearer goals to maintain confidence.

I mean, alright, if you say so. But, well, I'd rather hear this feedback coming from contributors, you know. It's a bit annoying to have people claiming to speak for you while they did not really experience the situation...

> Regardless I'm sorry, Juan. A game engine that developers depend on for successful projects requires sacrificing a bit of comfort so they understand the direction the project is taking and what they can rely on. That's the nature of devs interacting with your engine: their projects depend on it.

People probably need to re-read our license again: `THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.` You can't really expect a warranties without a contract.

**Comment by akien-mga** [comment]
So, to go back on my earlier post clarifying what **resources** the Godot project actually has.

1. Godot is one of the absolute largest community-developed open source project in the world. It's important to keep this in mind when considering what solutions known from e.g. AAA studios could be applicable. The comparison with Blender can be useful, but we're also very different projects (and Blender is 15 years old than Godot, much better funded and staffed, and much more mature overall - Godot is maturing rapidly, but shouldn't be expected to be at present Blender's level instantly).
2. We can't ask voluntary contributors to commit to a "comprehensive roadmap, public-facing task prioritization and schedule for the implementation of major features". That's just not something that any volunteer can commit to. They have lives, they're not being paid for this work, they can't commit to deadline like you would with a massive R&D team in a AAA studio. They may be busy with their dayjob, have shifting personal interests, get a kid, or simply a new hobby.
3. The only people we can order about to work on specific tasks outlined by management would be the handful of FTEs linked in our recent blog post (see my previous post). Half of those are the ones doing the management, so they can't also be the ones tasked with implementing prioritized tasks on a schedule. The rest are generalists who spend most of their time coordinating and reviewing the work on unpaid contributors. They also don't have that much time to focus 100% on their own feature or bugfix work. New work comes on their desk every day because unpredictable (non-hired) contributors open PRs in their areas every day. In other words, the vast majority of the hires of the foundation today are support functions for a community-developed engine. Refer to my first point. This is a unique situation from having a massive project developed majorly by volunteers in the open.
4. I have no objection to have a document that would outline some general goals/priorities that the core team identified, and plans to work on in the future, as time allows. It won't have deadline, it won't have strict prioritization. It won't go in the granularity level of identifying when a specific bug in a specific feature like SDFGI will be addressed, unless it's absolutely a showstopper that all Godot users keep complaining about (one random current example that we want to prioritize: bugs about the editor reporting scenes being corrupted after renaming files when using preloading and `class_name` due to the GDScript cache being out of sync).
5. We already have a production team dedicated to managing the project, with the unique perspective we have that most of the people we manage don't actually report to us because they're volunteer contributors working on their own time (and quite often during our off-hours, which is another challenge). So "hire a project manager" isn't really a solution, we've already done that and we're even in the process of hiring one more person to focus on production tasks. But production/project management can't define priorities and a roadmap for teams made solely of volunteer contributors. We can only surface issues to them, make clear that they seem high priority from our perspective, and hope they'll commit to fix them - or give them to one of our generalist hires to focus on, even if it's not their area of expertise. And well, this is *exactly* what we're doing.

**Comment by WickedInsignia** [comment]
@reduz I'm not proposing anyone should expect anything for free, and I don't believe anyone is.
Many contributors donate financially or with their time. You benefit directly from that financial assistance, as does your team. People provide that with the expectation the engine is improving in the ways that they deem important to its success and their uses.
If they don't believe that and are unsure of its direction, that flow of cash and time is reduced.


You keep focusing on this "social pressure" element.
If you're prioritisation results in that, you're doing it wrong. My point is not to do it wrong.
I don't believe going overly organic to avoid that entirely is the answer though, and it creates many problems in other areas.

I've already outlined in detail where the above strategies could greatly help users and contributors. Rather than any of it being considered, my proposal has been pushed back on in its entirety.

I don't believe what I've presented is of absolutely no value and no change is needed. There are obviously pressures and stresses with Godot's development that could be helped to some level by what I've mentioned above.

**Comment by WickedInsignia** [comment]
@akien-mga As I mentioned I'm not clueless as to the nature of FOSS projects and the way users and contributors interact with them.

I may be in the AAA industry but I find it extremely frustrating that throughout I've been categorised as representing it and all of its values in some way.
I don't think attributing any of what I've proposed to the nature of the AAA/corporate world is smart, or that none of it applies to FOSS at all.
These are simply sound project management principles that I believe apply to Godot in its current state, they don't turn an open and friendly project into a corporate hellhole. They're designed to better manage and communicate a large-scale project of Godot's nature, presented with the complexities of that project in mind.

**Comment by reduz** [comment]
@WickedInsignia 

As I wrote earlier, I don't currently benefit myself financially from donations (I am no longer invoicing the foundation), I get paid by my own company, so don't expect me to do work based on money donated. I am the one donating my work to Godot out of goodwill, and so is my company. 

If you get angry because I do some work on rendering that you don't understand what it is for (even though it is to fix exactly the issue you are complaining about), then be happy to know this and that donation money is not being misused.

With that said, If you donate to Godot expecting your bug being fixed, you are out of luck. Donations are intended to be used for what benefits the majority (which is generally the most pressing issues) and for project management, which is a monumental task. Donations do not happen in exchange of your specific bugs being fixed. As I said, your money is better off hiring someone else for this task. Those who donate already understand this, we have published several articles about this. It is also common sense.

**Comment by WickedInsignia** [comment]
Your company draws funding through the goodwill of Godot's development. The assertion that you don't financially benefit from your relation to Godot's development is disingenuous. W4 receives funding in large part due to its relationship to Godot.
You may not technically be paid by the foundation, but you are paid via the fact that Godot gains interest and development.

I've never at any point in this proposal or discussion demanded that donation money goes towards resolving my individual issues, either. Only that it is related to user and contributor confidence in the engine's development, and as such is subject to anything that might erode that confidence.

**Comment by akien-mga** [comment]
> I may be in the AAA industry but I find it extremely frustrating that throughout I've been somewhat characterised as representing it and all of its values in some way.

Fair enough, I apologize for taking this shortcut.

But you're only reacting to my first sentence on my first point. That doesn't invalidate the rest of my two posts, which IMO clearly explain why the solutions you outline, while they communicate a good intention and long term ideal, are simply not feasible in the Godot project as it is today.

I think I've explained clearly why we can't achieve a lot of what you're asking for with the resources we currently have. And what we can achieve, we already do.

**Comment by reduz** [comment]
@WickedInsignia 

> You may not technically be paid by the foundation, but you are paid via the fact that Godot gains interest and development.

This point really is out of place. Godot is an open ecosystem. A lot of people contributes and a lot of people benefit. This is not a zero sum game and this is not how FOSS works. Nobody is entitled to anything due to their contribution (be it money or code), and at the same time everyone benefits enormously from each others contribution. 

There is not a power game here based on who puts the most money because then if Google comes and puts a million dollars, then they can claim they want to be in control of the project.

**Comment by WickedInsignia** [comment]
@reduz None of that invalidates my reply nor that quote. It still stands as a fact that you financially benefit from your relationship with Godot's development.
I never said anything about Godot not being an open ecosystem, or that nobody else benefits, or that other people do not benefit from contributions, or that it is a power game, or that your financial benefit grants you power over the project. Those points are unrelated to the fact I was stating.

**Comment by WickedInsignia** [comment]
@akien-mga If you're confident that nothing can be changed or improved by what I've suggested there, and there is no intent to consider it, then I guess that settles it.

I've argued my points exhaustively and do not personally believe that no change can be made nor that Godot's development can continue in its current state without benefit from what I've suggested. I've considered all your points, and I still believe that.

Regardless, that's not enough to implement change or even attract discussion in good faith. I don't believe all parties here have argued with that in mind, and many characterisations or assumptions have been levelled for the convenience of serving their own convictions. I'm tired of fending those off and it's obvious from other interactions with individuals here that I'm not generally taken in good favour.

Thank you all for your time.

**Comment by reduz** [comment]
@WickedInsignia 

> @reduz None of that invalidates my reply nor that quote. It still stands as a fact that you financially benefit from your relationship with Godot's development.

Even if so, it does not entitle you (or anyone else) to ask me to fix your bug, to tell me what I should be working on, or putting timelines or anything like that. I gave you that answer so you understand that I am not currently being paid to fix your or other people's issues.

**Comment by WickedInsignia** [comment]
@akien-mga I understand the current state of Godot and that this could not be implemented right now, these are things to consider for the near future. I'm not sure why the assumption that my proposal comes with the demand it should happen now keeps occurring? I believe a lack of these tools is affecting users right now, but that doesn't mean I expect this all to be implemented right now.

As I've mentioned throughout, these tools could greatly benefit contributors and users. I don't believe that's invalidated by the fact they can't be implemented today, or the fact that some semblance of them already exists in a very dissonant form.
I don't believe where organisation and management in Godot is right now is the best it could be. I think you would agree with that.
It may be what's _possible_ now, but could be improved in the ways I've suggested when that becomes possible.


@reduz You are seeing demands that are not there.

I only summarized the nature of donations. You can believe what you like about entitlement: if Godot is not a promising project with confidence in its development and the good it could do for users, it suffers reduced interest and funding.
Would you be donating to a project you didn't think had potential? Probably not.

You may not be paid to fix other people's issues, but fixing other people's issues (along with reassuring them through good communication and prioritisation) sure does help you get paid.

**Comment by reduz** [comment]
@WickedInsignia 

> You may not be paid to fix other people's issues, but fixing other people's issues (along with reassuring them through good communication and prioritisation) sure does help you get paid.

You are going on a tangent here just to be "right". I never said I am not fixing other people's issues. I am doing precisely that. I only said (I repeat) **it does not entitle you (or anyone else) to ask me to fix your bug, to tell me what I should be working on, or putting timelines or anything like that**. 

Just like most people contributing to the project, I do it in my own time, at my own pace, without necessarily putting any roadmap or timelines. This is not going to change either for me or anyone else.

Seriously, I appreciate your views and what you do in the Godot community, there is no need for all this toxicity.

**Comment by WickedInsignia** [comment]
@reduz I'm defending what I said from mischaracterisation. I have no intentions to be wilfully obtuse.

I haven't demanded anything, nor suggested anything myself or anyone else is entitled to. I outlined the nature of donations and why people contribute, and how these suggestions around prioritisation and roadmapping could help assure user confidence is retained to safeguard that inflow.
People do not donate for you to do whatever you want. That's not to say you can't: I could sit and draw silly faces all day and not attract a single dollar because no one wants them.
If people feel their donations are going towards somebody else's time-wasting, rather than development for a tool that they and others could benefit from, they will be less inclined to donate.

Without a solid reference point for development, Godot can sometimes appear to be time-wasting.
You can do whatever you want, Juan. You can pretend donations aren't related to the development of Godot and spend those donations on things you want to do.
However, if users catch on that no important developments are being made and that their donations aren't leading to a better engine over time, they will be less inclined to keep donating and contributing.
I'm not saying that's what you're doing: that's an example to illustrate what I mean by the nature of donations and contributions.
I don't decide how that works, that's just the nature of it. I'm simply pointing it out.

You patently do not appreciate my views if you believe there's any toxicity in that. I've done nothing throughout this whole thread but clarify and argue for my suggestions while mitigating misreadings or misrepresentation of what I'm trying to say.

**Comment by yosoyfreeman** [comment]
Hi. I been using Godot for a while now, and saw it grow and mature. I consider myself a contributor, as i spend as much time and effort as i can finding bugs, testing features, reporting issues and participating in UX discussions, so i would like to make some quick notes.

-I believe this conversation is extremely useful and that it is slowly going towards a defensive behaviour from all parts, which could prevent a really healthy conversation. I perfectly understand the pressure the Godot core contributors feels, new people may not be totally aware that they are arriving in a point of exponential growth, in which not so far ago Godot was a much simpler piece of software without all the support and contributors they now have. Is a transition point, and they are doing their best to accommodate to the expectations people are putting on the software. We should have in mind that, during this time, really hard accusations has been made towards them. On the other Side, i myself i'm ashamed to say i had a misunderstanding with @WickedInsignia because their assertive tone was misinterpreted by me as a lack of good faith of realism, and the other way around. During this time tho, i seen that they worked really hard to find issues, and not only that, but tried to get as close the the source of the bugs as possible. I can understand the frustration seeing that all that work **_seems_** to not be fruitful. With this i just want to say that Godot being the big project it is, and the fact that is kind of hard to know what is exactly working on (At least from the outside) can cause situation in which may appear like a "side" is being unreasonable with the other, while in reality they are just different perspectives. The Godot core team is doing is job, which is taking care of the engine itself and implementing stuff. Insignia is an excellent professional and their job as art director is to find issues and propose solutions. In my eyes, that and this conversation itself is contributing too.

-I agree that maybe not a roadmap, but a clear way of seeing a list of issues on core elements in which users can find essential things to fix may be good. Not to pressure nobody, but because keeping track with the github is a full time job itself. This may help user find task that they can solve and didn't knew they needed to be solved.

-I would love to have more documentation and official videos about the process of contributing to Godot. Lots of us come from making games, not contributing to C++ engine source code. Even tho, with a better guidance, some of us would love to be able to make the step and start working with the simple stuff. I think having more educational content about how and why things are done the way they are done could contribute to fix this "disconnection" between the team and the community.

Please, you all here are extremely talented people working for the same goal. Keep that in mind in the heat of the conversation.

**Comment by WickedInsignia** [comment]
Thanks @yosoyfreeman , I think you've put it all very succinctly there.

You're correct, what I'm expressing can definitely come across as assertive. A little of that is (unfortunately) required in FOSS sometimes due to the nature of open development. Sometimes you need to push your idea through the gauntlet of doubt and stand up for your proposals. Sometimes, that push is too firm.
All of the above is provided as suggestion, and all of the discussion on my side is in defense of its validity.

I want Godot to be the best it can be. That goes hand-in-hand with being well-managed, accessible and retaining of user confidence. I've tried my best to identify where I think some of that is lacking, based on both my own experiences and in discussion with other parties, along with my professional experiences.
Those suggestions may not be useful in their entirety: I expect that. I would hope they're at least considered. I get the feeling that the core team believes I misunderstand the project: that may be partially true, but I don't believe I misunderstand it so severely that nothing I've said could be applied or considered at some point. I'm not new to large-scale project development nor FOSS, and I acknowledge that few here are not experienced with either.

I hope this spurs discussion and helps inform future plans in some way. If it's determined that they will not help after due consideration, then I'll be glad they were at least impartially assessed.

**Comment by reduz** [comment]
@WickedInsignia To me courtesy is taking the time to answer you because I value you instead of ignoring you, if to you courtesy is only agreeing with you and your proposal then we are not on the same channel.

**Comment by WickedInsignia** [comment]
I never said that, and again you are characterizing. There is nothing here to signify people should wholly agree with my proposal, you are inventing that to antagonize my intentions and paint the picture that I am completely uncompromising. I've seen you do this to others and I hoped it wouldn't come to that.

You're a real piece of work man.
Best to the team.

**Comment by robbertzzz** [comment]
What I'm seeing in the original OP and got lost in the discussion is a lack of direction in the engine's development. A roadmap might be the wrong term, but the idea of sharing what the production team wants to focus on next would be hugely beneficial.

Imagine that, instead of all areas equally improving, the next second-digit release (4.3 at this time) would have a big focus on tilemap editing. This has been a major pain point on various social media where users are both happy and frustrated by the new (as of 4.0) implementation of a tilemap system. _Instead of telling contributors what to work on_, having this kind of focus on one feature for a next minor release would help set expectations for end users, it would show contributors why their work isn't being prioritised, and it would help the production team focus on specific pull requests rather than being all over the place. This fictitious 4.3 release wouldn't _just_ include improved tilemap editing as bug fixes and small improvements are always necessary, but this focus can motivate contributors to work on and discuss a specific area, it can set goals for what people want out of the 4.3 release and it provides a clear improvement over the previous engine version that people can point to. It also provides something tangible that people can test, hopefully resulting in a better feedback loop between core team and end users.

You core team members are all absolutely right that you can't force contributors to work on a specific thing, but you can absolutely control which things get your attention, which things get the public's attention, and which PRs get merged. I think Blender is a great example here, where every release has a specific focus (from memory I've seen an Eevee-specific release and an "everything nodes" project spanning multiple releases, each focusing on one part being node-ified).

**Comment by WickedInsignia** [comment]
To summarize, the proposal came down to:
- A public easy-reference roadmap to list intended long-term goals where possible
- Public prioritization for bugs/new features so energy is focused on what devs actually need
- A non-concrete schedule with expected completion dates for features and releases

The disclaimer was made continuously that none of this is binding and free contributors wouldn't be expected to commit to work they don't want to do. This would help contributors who do want to contribute to the hard but necessary work, and help devs/users understand why certain features are being prioritized before others.
It's been a point of contention for some time, and in my discussions with professional devs the lack of direction is a major reason they hesitate to rely on Godot. Therefore I thought it best to propose and discuss.

I admit to being very assertive throughout. That came from a desire to help alleviate the issues and retain Godot's public rep, but I understand why it led to pushback in turn. I regret the initial tone and how this went down, but I don't apologize for bringing it up. I'm confident that my experiences with the engine and other game/engine devs justified the appraisal.
To my understanding there's some internal conflict on this topic and it's not clear what's being worked on, but I am not pursuing it further.

This is probably better left to internal discussion and not reopened.

**Comment by robbertzzz** [comment]
> This is probably better left to internal discussion and not reopened.

I get where you're coming from, but I felt that the point I found the most important was missed in this discussion due to things getting heated. You mentioned it, but nobody really seemed to respond to it in a constructive way.

I think it's good to voice concerns from a professional's view and I think discussing those in a public place makes sense for an open project like Godot. As a fellow professional within the industry I agree with most of your points and I think it's worth mentioning that to the team. The boom that Blender has gone through is, in my opinion, largely due to project management focusing on a usable tool rather than the mess of unfinished features that it was before. I feel like Godot is in a similar place now where it's mature enough to rethink how the project is managed in order to bring it to a next level.

I see this kind of discussion as a huge compliment to the contributors and core developers. It means that their work is being seriously considered as a viable tool by industry professionals and in your case this notably includes AAA teams. This basically is an entirely new target audience entering the chat with completely different wishes from a project like this. The problems they see aren't the actual features anymore (we all know that AA and AAA teams commonly have in-house versions of source available engines like Unreal), but instead they are about expectations. They need to know they can rely on this engine for their long-term projects.

**Comment by WickedInsignia** [comment]
> I get where you're coming from, but I felt that the point I found the most important was missed in this discussion due to things getting heated. You mentioned it, but nobody really seemed to respond to it in a constructive way.

Ha yup. I know why but still, I can't do much now if the delivery didn't land. That's my bad.
The pushback here and public discussions in the Discord seem to indicate the team would rather keep these discussions internal and an external third party had no place proposing this. Which is fine, but I don't think many people will recognize the boxed responses and thinly-veiled dismissals here that have recurred throughout Godot's recent history when similar discussions come up. I don't think the team realizes how this appears on the broad scale.

Godot's messaging needs a lot of work. One day it's nearly AA/AAA ready, the next it's not. One day it's comparable to Unity and plugin creators are silly not to flock to it, the next day it's unfair to compare it to Unity. Godot is extremely powerful with high-quality code and a strong community, but please don't dunk on us because we're very little and helpless. They will vaguely dunk on other projects, but any return fire is uncalled for.
Professionals see this, talk about it behind closed doors, and are not impressed. I can personally attest to that.

This roadmapping issue has been brought up before and will be brought up again. I'm not worried if everything didn't land now, similar discussions will pester Godot's development (especially from experienced parties) until it's resolved to a satisfactory point. I won't be continuing with it though and did my (admittedly somewhat ungraceful) part, so I'm considering this proposal closed.

**Comment by WickedInsignia** [comment]
> I get where you're coming from, but I felt that the point I found the most important was missed in this discussion due to things getting heated. You mentioned it, but nobody really seemed to respond to it in a constructive way.

My tone wasn't terribly inviting, it was very direct in a way that maybe needed to be heard but prompted defensiveness. That's my bad. I agree with everything you've said there though.

Godot's messaging needs a lot of work. One day it's nearly AA/AAA ready, the next it's not. One day it's comparable to Unity and plugin creators are silly not to flock to it, the next day it's unfair to compare it to Unity. Godot is extremely powerful with high-quality code and a strong community, but please don't dunk on us because we're very little and helpless. They will vaguely dunk on other projects, but any return fire is uncalled for.
Professionals see this, talk about it behind closed doors, and are not impressed. I can personally attest to that.

The core team needs to stop selling people on the idea that Godot is the future and is ready, then put up defenses when they're criticized for basic shortcomings. When people go looking for more info they find no roadmap or long term goals and no prioritization, but plenty of excuses.
You can't excuse away bad managerial practice. People see that stuff. Maybe the leads have convinced themselves they're well-managed, but this is a FOSS engine. **People know when it's badly managed and poorly built.** The proof is in the pudding.
Hobbyists and enthusiasts won't pick up on that but experienced devs will and they do. If you're lucky they'll call it out but most of the time they'll just walk away. No one has time to argue with a defensive FOSS engine that takes shots at other people's work.

This roadmapping issue has been brought up before and will be brought up again. I'm not worried if everything didn't land now, similar discussions will pester Godot's development (especially from experienced parties) until it's resolved to a satisfactory point. I won't be continuing with it though and did my (admittedly somewhat ungraceful) part, so I'm considering this proposal closed.

**Comment by robbertzzz** [comment]
> I won't be continuing with it though and did my (admittedly somewhat ungraceful) part, so I'm considering this proposal closed.

You don't need to actively engage in this conversation for it to stay open, feel free to turn off notifications while the discussion continues.
